Editorial review: Document AudioPlaybackStats (#43719)
mdn/content: files/en-us/web/api/audioplaybackstats/index.md

---
title: AudioPlaybackStats
slug: Web/API/AudioPlaybackStats
page-type: web-api-interface
browser-compat: api.AudioPlaybackStats
---

{{APIRef("Web Audio API")}}

> The **`AudioPlaybackStats`** interface of the [Web Audio API](/en-US/docs/Web/API/Web_Audio_API) provides access to duration, underrun, and latency statistics for the associated {{domxref("AudioContext")}}. These statistics allow you to measure audio delay and glitchiness.

Reviewer: glitchiness -> glitches?

Author: done

> The **`AudioPlaybackStats`** interface of the [Web Audio API](/en-US/docs/Web/API/Web_Audio_API) provides access to duration, underrun, and latency statistics for the associated {{domxref("AudioContext")}}. These statistics allow developers to measure audio delay and glitchiness.

Reviewer: developers = you throughout?

Author: Agreed. Updated.

> An audio context's `AudioPlaybackStats` object can be accessed via its {{domxref("AudioContext.playbackStats")}} property. The returned `AudioPlaybackStats` object is live — the contained property values are updated once every second.

Reviewer: "You can access an audio context's ..."

## Description

> When audio is playing in an application, it is beneficial to measure {{glossary("latency")}} and **underrun** — both can contribute to a bad user experience:

Reviewer: ```suggestion
When audio is playing in an application, it is beneficial to measure audio {{glossary("latency")}} and **underrun** because both can contribute to a bad user experience:
```
- we could maybe unbold underrun here and instead add emphasis to the terms on lines 18 and 20
- "When audio is playing in an application" -> how about "In applications that play audio" or "Audio applications can benefit from measuring..."?
- "bad user experience" -> how about we make this specific as "poor audio experience"?

Author: I liked all these suggestions, so I've applied them all.

- **Audio latency**
  - : A measure of the delay between the user activating a control (such as a play button) and the audio playing as expected. Significant latency can make an application feel unresponsive.
- **Underrun**

>   - : A gap in playback when the audio application runs out of buffered audio data to play before new data arrives to replace it — in other words, it can't provide audio frames to the output device fast enough. This can happen due to audio graph complexity, CPU overload, or malfunctions in other audio programs. The result is an audible "glitch" — a click, a pop, or an audio dropout — because the application has nothing to play and fills the gap with silence or noise.

Reviewer: "audio application": should this be audio context?

>   - : Happens when the audio application runs out of buffered audio data to play before new data arrives to replace it — in other words, it can't provide audio frames fast enough. This occurs when audio is not delivered to the playback device in time due to audio graph complexity, CPU overload, or malfunctions in other audio programs. The result is an audible "glitch" — a click, pop, or audio dropout occurring because the application has nothing to play and fills the gap with silence or noise.

Reviewer: - can we qualify "it" in the first sentence?
```suggestion
  - : A gap in playback when the audio application runs out of buffered audio data to play before new data arrives to replace it — in other words, it can't provide audio frames to the output device fast enough. This can happen due to audio graph complexity, CPU overload, or malfunctions in other audio programs. The result is an audible "glitch" — a click, a pop, or an audio dropout — because the application has nothing to play and fills the gap with silence or noise.
```

Reviewer: should it be "...when the audio context runs.."

Author: I think the suggestion works really well as-is. Applied in the next commit.

> If you detect underruns, take action to avoid more in future – for example, by providing a larger buffer or by freeing up system resources. Use a larger buffer carefully because they can increase latency; it is important to achieve a balance. You can reduce latency by simplifying the required processing or by reducing the playback buffer size.

Reviewer: - "it" for buffer
- I guess we can drop the "balance" part, it is sort of implied
```suggestion
If you detect underruns, take action to avoid more in future – for example, by providing a larger buffer or by freeing up system resources. Use a larger buffer carefully because it will also increase latency. You can reduce latency by simplifying the required processing or by reducing the playback buffer size.
```

> If underruns are detected, action should be taken to avoid future underruns, for example providing a larger buffer, or freeing up system resources. Larger buffers should be used carefully as they can also increase latency, so it is important to achieve a balance. Latency can be reduced by simplifying the required processing, or reducing the playback buffer size.

Reviewer: basically switching all to active voice
```suggestion
If you detect underruns, take action to avoid more in future – for example, by providing a larger buffer or by freeing up system resources. Use a larger buffer carefully because they too can increase latency, so it is important to achieve a balance. You can reduce latency by simplifying the required processing or by reducing the playback buffer size.
```

Author: Used, but tweaked a little bit.

> Web audio performance varies greatly across devices, from high-end modern desktop computers to low-end budget mobile phones. The `AudioPlaybackStats` object allows you to collect telemetry data from your users to understand how your app performs in the "real world". Use this data to identify and respond to latency and underrun problems.

Reviewer: greatly -> widely?

> Web audio performance varies greatly between high-end devices such as a modern desktop computer and low-end devices such as budget mobile phones. The `AudioPlaybackStats` object allows developers to collect telemetry data from their users to understand how their app performs in the "real world" with respect to latency and underrun and respond to problems as they arise.

Reviewer: 2nd sentence can be split into two
```suggestion
Web audio performance varies greatly across devices, from high-end modern desktop computers to low-end budget mobile phones. The `AudioPlaybackStats` object allows you to collect telemetry data from your users to understand how your app performs in the "real world". Use this data to identify and respond to latency and underrun problems.
```

> For example, you could create an "adaptive" audio system in your app that detects when underrun or latency exceeds a certain threshold (where the audio starts glitching), and:

Reviewer: - could also end the lead-in with "...and take the following actions:"
```suggestion
For example, you could create an "adaptive" audio system that detects when underrun or latency exceeds a certain threshold (when the audio starts glitching), and:
```

Author: Applied in my next commit. It needed the extra lead-in bit for the structure to work with your subsequent suggestion, but it works fine now, altogether.

- Reduce computation load by cutting down the maximum number of voices in play at once or removing complex filters.

> - Prompt the user to close other tabs or apps or to switch their audio output device.

Reviewer: sorry, with the current lead-in, the verb forms you had used originally are correct: Reduces and Prompts

> - Prompts the user to try closing other tabs or apps or switching their audio output device for better performance.

Reviewer: ```suggestion
- Reduce computation load by cutting down the maximum number of voices in play at once or removing complex filters.
- Prompt the user to close other tabs or apps or to switch their audio output device.
```

Author: This works nicely. I wanted to look at it in context locally; I've applied it in my next commit.

> ### Provided underrun stats

Reviewer: Not sure what this and the next heading means.
Is it along these lines: "Underrun statistics provided by the interface"?

Author: I've made this change, so this heading is consistent with the next heading after I applied your suggestion there.

Reviewer: Thanks

> Underruns are defined in terms of **underrun frames** and **underrun events**:

Reviewer: or "The interface reports underrun data using ..."

Author: I'm not so sure about this change, because that implies that underrun frames and underrun events are particular to this API, whereas they are wider concepts in audio in general.

Reviewer: Fair point.

How about using "expressed" instead of "defined"?

Also like in the DL earlier, how about moving emphasis to the terms in the DL instead of in line 34?

- Underrun frame

>   - : An audio frame, typically silence in the case of a web application, played by the output device in the absence of any "real" audio data from the audio context.

Reviewer: ```suggestion
  - : An audio frame, typically silence in the case of a web application, played by the output device when there is no actual audio data from the audio context.
```

- Underrun event
  - : The playback of a continuous sequence of underrun frames. The duration of the underrun event is the total duration of the sequence of underrun frames.

> The number of underrun events that have occurred since the audio context was first initialized is reported by the {{domxref("AudioPlaybackStats.underrunEvents")}} property, and the duration of those underrun events is reported by the {{domxref("AudioPlaybackStats.underrunDuration")}} property. This allows developers to find out exactly how often and for how long audio is cutting out due to underrun.

Reviewer: ```suggestion
The number of underrun events since the audio context initialization is reported by the {{domxref("AudioPlaybackStats.underrunEvents")}} property, and the duration of those underrun events is reported by the {{domxref("AudioPlaybackStats.underrunDuration")}} property. This allows you to find out how often and for how long the audio cuts out due to underrun.
```

> ### Provided latency stats

Reviewer: this?
```suggestion
### Latency statistics provided by the interface
```

The audio context's latency can be measured using the {{domxref("AudioPlaybackStats.averageLatency")}}, {{domxref("AudioPlaybackStats.minimumLatency")}}, and {{domxref("AudioPlaybackStats.maximumLatency")}} properties.

> It is possible to retrieve the immediate playout latency of the context via the {{domxref("AudioContext.outputLatency")}} property, however this is an instantaneous value that fluctuates rapidly. `AudioPlaybackStats` provides average, minimum, and maximum latency over time, which is much more useful for identifying persistent performance issues.

Reviewer: ```suggestion
It is possible to retrieve the immediate playout latency of the audio context via the {{domxref("AudioContext.outputLatency")}} property; however, this is an instantaneous value that fluctuates rapidly. `AudioPlaybackStats` provides the average, minimum, and maximum latency over time, which is more useful for identifying persistent performance issues.
```

## Instance properties

- {{domxref("AudioPlaybackStats.averageLatency")}} {{ReadOnlyInline}}

>   - : A number indicating the average latency since the audio context was first initialized, or since {{domxref("AudioPlaybackStats.resetLatency()")}} was last called.

Reviewer: - "first" before initialized and comma before "or since" can be dropped from all property descriptions

Author: done, here and in the property detail pages.

Reviewer: Awesome

- {{domxref("AudioPlaybackStats.minimumLatency")}} {{ReadOnlyInline}}
  - : A number indicating the minimum latency since the audio context was first initialized, or since {{domxref("AudioPlaybackStats.resetLatency()")}} was last called.
- {{domxref("AudioPlaybackStats.maximumLatency")}} {{ReadOnlyInline}}
  - : A number indicating the maximum latency since the audio context was first initialized, or since {{domxref("AudioPlaybackStats.resetLatency()")}} was last called.
- {{domxref("AudioPlaybackStats.totalDuration")}} {{ReadOnlyInline}}
  - : A number indicating the total duration of all audio frames since the audio context was first initialized.
- {{domxref("AudioPlaybackStats.underrunDuration")}} {{ReadOnlyInline}}

>   - : A number indicating the total duration of the underrun events that have occurred since the audio context was first initialized.

Reviewer: ```suggestion
  - : A number indicating the total duration of the underrun events since the audio context was initialized.
```

- {{domxref("AudioPlaybackStats.underrunEvents")}} {{ReadOnlyInline}}

>   - : A number indicating the number of underrun events since the audio context was initialized.

Reviewer: just spotted the "number...number". We could change it to "An integer...number of..."

>   - : A number indicating the number of underrun events that have occurred since the audio context was first initialized.

Reviewer: ```suggestion
  - : A number indicating the number of underrun events since the audio context was initialized.
```

## Instance methods

- {{domxref("AudioPlaybackStats.resetLatency()")}}

>   - : Sets the interval over which the latency stats are measured to the current time onwards.

Reviewer: This was slightly unclear, so I had to check the spec. Might be good to rephrase it to "Sets the start of the interval" instead of "Sets the interval":

```suggestion
  - : Sets the start of the interval during which latency statistics are measured to the current time.
```
or maybe to highlight the "reset" feature:

```suggestion
  - : Resets the start of the interval during which latency statistics are measured to the current time.
```

or a simpler: Resets the latency measurement window to start from the current time

WDYT?

Author: I went for the second one, as I like the way it highlights the "reset". Updated in both places.

- {{domxref("AudioPlaybackStats.toJSON()")}}
  - : A {{Glossary("Serialization","serializer")}} that returns a JSON representation of the {{domxref("AudioPlaybackStats")}} object.

## Examples

> ### Reporting audio playback stats

Reviewer: You might want to search for "stats" and replace the instances you like with "statistics"

This example demonstrates how to report audio stats accessed via the {{domxref("AudioPlaybackStats")}} object.

#### HTML

We include three {{htmlelement("button")}} elements — one to start the audio playing, one to retrieve and display a set of stats, and one to run the {{domxref("AudioPlaybackStats.resetLatency()")}} method. We also include a {{htmlelement("ul")}} element inside which stats will be displayed.

```html live-sample___playback-stats
<p>
  <button class="play">Play audio</button>
  <button class="stats">Display stats</button>
  <button class="reset">Reset latency</button>
</p>
<hr />
<ul class="output"></ul>
```

```css hidden live-sample___playback-stats
ul {
  width: 80%;
  margin: 0 auto;
}
li {
  margin-bottom: 10px;
}
```

#### JavaScript

> In our JavaScript, we begin by grabbing references to the buttons and the output list. We also disable our stats and reset buttons, because they won't do anything initially. We'll enable them again once they have event listeners attached.

Reviewer: ```suggestion
In our JavaScript, we begin by grabbing references to the buttons and the output list. We also disable the stats and reset buttons because they won't do anything initially. We'll enable them again once they have event listeners attached.
```

```js live-sample___playback-stats
const playBtn = document.querySelector(".play");
const statsBtn = document.querySelector(".stats");
const resetBtn = document.querySelector(".reset");
const output = document.querySelector(".output");

statsBtn.disabled = true;
resetBtn.disabled = true;
```

> Next, we add a `click` event listener to the play button so that when it is clicked, we:

Reviewer: - We'll need to match all other instances of buttons to their labels in the HTML
- Also for the UI label, I think the convention is to use bold emphasis, so we can drop quotes around button names
```suggestion
Next, we add a `click` event listener to the **Play audio** button so that when the user clicks it, we:
```

- Create a new {{domxref("AudioContext")}} and disable the play button so it can't be pressed again.

> - Run some feature detection code that checks if the {{domxref("AudioContext.playbackStats")}} property exists. If it doesn't, we display a "Your browser doesn't support AudioPlaybackStats." message in a list item in the output list, and `return` out of the function.

Reviewer: ```suggestion
- Run some feature detection code that checks if the {{domxref("AudioContext.playbackStats")}} property exists. If it doesn't, we display the "Your browser doesn't support AudioPlaybackStats." message in the output list and `return` from the function.
```

> - Create a basic audio graph consisting of of an {{domxref("OscillatorNode")}} and a {{domxref("GainNode")}} and start the oscillator playing.

Reviewer: ```suggestion
- Create a basic audio graph consisting of an {{domxref("OscillatorNode")}} and a {{domxref("GainNode")}}, and start the oscillator.
```

> - Enable the stats button and give it a `click` event listener so that when it is clicked, we write the different stats available in the audio context's {{domxref("AudioPlaybackStats")}} object into a text string and display it in a list item in the output list.

Reviewer: or the shorter
```suggestion
- Enable the stats button and give it a `click` event listener so that when it is clicked, we write the statistics from the audio context's {{domxref("AudioPlaybackStats")}} object and display them in the output list.
```

    output.appendChild(listItem);
    return;
  }

  const oscillator = audioCtx.createOscillator();
  oscillator.type = "square";
  oscillator.frequency.setValueAtTime(100, audioCtx.currentTime);
  const gain = audioCtx.createGain();
  gain.gain.value = 0.006;

  oscillator.connect(gain);
  gain.connect(audioCtx.destination);
  oscillator.start();

  const stats = audioCtx.playbackStats;

  statsBtn.disabled = false;
  statsBtn.addEventListener("click", () => {
    const listItem = document.createElement("li");
    const statsText = `Underrun duration: ${stats.underrunDuration}
                       Underrun events: ${stats.underrunEvents}
                       Total duration: ${stats.totalDuration}
                       Average latency: ${stats.averageLatency}
                       Min latency: ${stats.minimumLatency}
                       Max latency: ${stats.maximumLatency}`;
    listItem.textContent = statsText;
    output.appendChild(listItem);
  });

  resetBtn.disabled = false;
  resetBtn.addEventListener("click", () => {
    stats.resetLatency();
  });
});
```

#### Result

The rendered output looks like this:

{{embedlivesample("playback-stats", "100%", "400")}}

> Click the "Play audio" button to start the ocillator tone playing. Now when you click the "Display stats" button, you'll see the different stats available in the {{domxref("AudioPlaybackStats")}} object displayed in a list item.

Reviewer: ```suggestion
Click the "Play audio" button to start the oscillator tone playing. When you click the "Display stats" button, you'll see the statistics available in the {{domxref("AudioPlaybackStats")}} object.
```

> If you click the "Reset latency" button, then click the "Display stats" button, you'll see new stats displayed, but the minimum latency will no longer be zero. This is because the latency stats will now only be measured from the time you clicked the "Reset latency" button, not from the start of the audio graph.

Reviewer: ```suggestion
If you click the "Reset latency" button, and then click the "Display stats" button, new statistics will appear, but the minimum latency will no longer be zero. This is because the latency is now measured from when you clicked the "Reset latency" button, not from when the audio context was initialized.
```

Author: Applied in https://github.com/mdn/content/pull/44088
